plutus-tx/src/PlutusTx/Data/List.hs

```diff
@@ -0,0 +1,194 @@
+toSOP :: (UnsafeFromData a) => List a -> [a]
+toSOP (List l) = go l
+  where
+    go = B.caseList' [] (\h t -> unsafeFromBuiltinData h : go t)
+
+{-# INLINEABLE fromSOP #-}
+fromSOP :: (ToData a) => [a] -> List a
+fromSOP = List . BI.unsafeDataAsList . B.mkList . fmap toBuiltinData
+
+{-# INLINEABLE find #-}
+find :: (UnsafeFromData a) => (a -> Bool) -> List a -> Maybe a
+find pred (List l) = go l
+  where
+    go =
+        B.caseList'
+            Nothing
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in
+                    if pred h'
+                        then Just h'
+                        else go t
+            )
+
+-- TODO: return [] or List?
+{-# INLINEABLE findIndices #-}
+findIndices :: (UnsafeFromData a) => (a -> Bool) -> List a -> [Integer]
+findIndices pred (List l) = go 0 l
+  where
+    go i =
+        B.caseList'
+            []
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                    indices = go (B.addInteger i 1) t
+                in
+                    if pred h'
+                        then i : indices
+                        else indices
+            )
+
+-- TODO: return [] or List?
+{-# INLINEABLE filter #-}
+filter :: (UnsafeFromData a) => (a -> Bool) -> List a -> [a]
+filter pred (List l) = go l
+  where
+    go =
+        B.caseList'
+            []
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in if pred h' then h' : go t else go t
+            )
+
+-- TODO: return [] or List?
+{-# INLINEABLE mapMaybe #-}
+mapMaybe :: (UnsafeFromData a) => (a -> Maybe b) -> List a -> [b]
+mapMaybe f (List l) = go l
+  where
+    go =
+        B.caseList'
+            []
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in case f h' of
+                    Just b  -> b : go t
+                    Nothing -> go t
+            )
+
+{-# INLINEABLE any #-}
+any :: (UnsafeFromData a) => (a -> Bool) -> List a -> Bool
+any pred (List l) = go l
+  where
+    go =
+        B.caseList'
+            False
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in pred h' || go t
+            )
+
+{-# INLINEABLE foldMap #-}
+foldMap :: (UnsafeFromData a, Monoid m) => (a -> m) -> List a -> m
+foldMap f (List l) = go l
+  where
+    go =
+        B.caseList'
+            mempty
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in f h' <> go t
+            )
+{-# INLINEABLE map #-}
+map :: (UnsafeFromData a, ToData b) => (a -> b) -> List a -> List b
+map f (List l) = List (go l)
+  where
+    go =
+        B.caseList'
+            B.mkNil
+            (\h t -> BI.mkCons (toBuiltinData $ f $ unsafeFromBuiltinData h) (go t))
+
+{-# INLINEABLE length #-}
+length :: List a -> Integer
+length (List l) = go l 0
+  where
+    go =
+        B.caseList'
+            id
+            (\_ t -> B.addInteger 1 . go t)
```

Comment:
I wouldn't be surprised if that actually performs worse than the eta-expanded version.